src/results_analyze/rules_analyze.py

```diff
@@ -0,0 +1,299 @@
+#
+# You may not use this software except in compliance with the License.
+# You may obtain a copy of the License at: http://apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software distributed
+# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+# CONDITIONS OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the License.
+#
+# When you publish or redistribute any data created with ScanCode or any ScanCode
+# derivative work, you must accompany this data with the following acknowledgment:
+#
+#  Generated with ScanCode and provided on an "AS IS" BASIS, WITHOUT WARRANTIES
+#  OR CONDITIONS OF ANY KIND, either express or implied. No content created from
+#  ScanCode should be considered or used as legal advice. Consult an Attorney
+#  for any legal advice.
+#  ScanCode is a free software code scanning tool from nexB Inc. and others.
+#  Visit https://github.com/nexB/scancode-toolkit/ for support and download.
+
+import yaml
+import os
+# import itertools
+# import numpy as np
+import pandas as pd
+
+from results_analyze.df_file_io import DataFrameFileIO
+
+# Format in which the Information DataFrames will be stored in hdf5 files
+HDF5_STORE_FORMAT = 'fixed'
+# Threshold of Words which is used in `Rule.compute_relevance` in `scancode.licensedcode.models.py`
+THRESHOLD_COMPUTE_RELEVANCE = 18.0
+
+
+class LicenseRulesInfo:
+    def __init__(self, has_loaded=False):
+
+        self.df_io = DataFrameFileIO()
+
+        self.hdf5_info_dfs = os.path.join(os.path.dirname(__file__), 'data/lic-rule-info/info_dfs.hdf5')
+
+        self.rule_df = None
+        self.lic_df = None
+
+        self.key_dict = None
+        self.identifier_dict = None
+        self.key_df = None
+        self.identifier_df = None
+
+        self.rule_bool_cols = ["is_license_reference", "is_license_text", "is_license_notice", "is_license_tag",
+                               "only_known_words", "is_false_positive", "is_negative"]
+
+        if has_loaded:
+            self.load_info_from_file()
+        else:
+            self.lic_folder_path, self.rule_folder_path = self.df_io.load_folder_names()
+            self.load_info_from_scancode()
+
+    def load_info_from_file(self):
+        """
+        Load License and Rule DataFrames and Conversion Dictionary DataFrames from hdf5 files.
+        """
+
+        self.rule_df = self.df_io.load_dataframe_from_hdf5(self.hdf5_info_dfs, "rule")
+        self.lic_df = self.df_io.load_dataframe_from_hdf5(self.hdf5_info_dfs, "lic")
+        self.identifier_df = self.df_io.load_dataframe_from_hdf5(self.hdf5_info_dfs, "id")
+        self.key_df = self.df_io.load_dataframe_from_hdf5(self.hdf5_info_dfs, "key")
+
+        self.identifier_dict = self.df_io.df_to_inv_dict(self.identifier_df)
+        self.key_dict = self.df_io.df_to_inv_dict(self.key_df)
+
+    def save_info_to_file(self):
```

Comment:
More of a comment, but what does saving an intermediate file get you?

Reply:
So each time we're running Scan Results through this pipeline, we're loading License/Rule Information. Now there are two ways to do this obviously, 

1. From the Scancode Rules/Licenses Folders. Loading and Formatting Everything. This reads the `.yaml`, `.LICENSE` and `.RULE` files and formats it into DataFrames.
2. Load the DataFrames directly from Binary Files.

Obviously Option 1 requires more computation, we don't need to load from Scancode Files always, but only when Rules/Licenses are updated, and otherwise just take the input from the binary files.

Reply:
And loading the `.LICENSE` and `.RULE` files from the filesystem into a DataFrame causes some sort of index to happen, hence the computational load you mention?

Reply:
Some sort of yes, so in this case the information in files are processed and the DataFrame is created, and in the other case, an already created DataFrame is loaded from a Binary file.